### PR TITLE
dapper transactions

### PR DESCRIPTION
**Pull request by bjartek**

### CODE REVIEW

File: transactions/registerDapper.cdc [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		self.leases=account.borrow<&FIND.LeaseCollection>(from: FIND.LeaseStoragePath)
+	}
+
+	pre{
+		self.vaultRef != nil : "Could not borrow reference to the fusdVault!"
+		self.leases != nil : "Could not borrow reference to find lease collection"
+		self.price == amount : "Calculated cost : ".concat(self.price.toString()).concat(" does not match expected cost : ").concat(amount.toString())
+	}
+
+	execute{
```
Comment (bjartek) [inline transactions/registerDapper.cdc:24]: This transaction needs a post request to verify that duc is not leaked

File: contracts/FIND.cdc [hunk 2]
```diff
@@ -614,6 +622,48 @@ pub contract FIND {
...
+			let addonPrice = network.addonPrices[addon]!
+
+			let lease = self.borrow(name)
+
+			if lease.addons.containsKey(addon) {
+				panic("You already have this addon : ".concat(addon))
+			}
+
+			if vault.balance != addonPrice {
+				panic("Expect ".concat(addonPrice.toString()).concat(" FUSD for ").concat(addon).concat(" addon"))
```
Comment (bjartek) [inline contracts/FIND.cdc:650]: this says FUSD but it should be Dapper Credit

File: contracts/FIND.cdc [hunk 3]
```diff
@@ -1319,6 +1387,36 @@ pub contract FIND {
...
+					//the name is taken but not expired so we extend the total period of the lease
+					lease.setValidUntil(lease.validUntil + self.leasePeriod)
+				} else {
+					lease.setValidUntil(Clock.time() + self.leasePeriod)
+				}
+				lease.setLockedUntil(lease.validUntil+ self.lockPeriod)
+
+				let cost= self.calculateCost(name)
+				if vault.balance != cost {
+					panic("Vault did not contain ".concat(cost.toString()).concat(" amount of FUSD"))
```
Comment (bjartek) [inline contracts/FIND.cdc:1407]: FUSD > Dapper Credit

File: contracts/FIND.cdc [hunk 4]
```diff
@@ -1370,6 +1468,36 @@ pub contract FIND {
...
+			}
+
+			//if we have a locked profile that is not owned by the same identity then panic
+			if nameStatus.status == LeaseStatus.LOCKED {
+				panic("Name is locked")
+			}
+
+			let cost= self.calculateCost(name)
+			if vault.balance != cost {
+				panic("Vault did not contain ".concat(cost.toString()).concat(" amount of FUSD"))
```
Comment (bjartek) [inline contracts/FIND.cdc:1491]: FUSD > Dapper Credit

File: contracts/FIND.cdc [hunk 5]
```diff
@@ -1741,6 +1869,28 @@ pub contract FIND {
 
 	}
 
+	access(contract) fun checkMerchantAddress(_ merchAccount: Address) {
+		// If only find can sign the trxns and call this function, then we do not have to check the address passed in.  
+		// Otherwise, would it be wiser if we hard code the address here? 
+
+		if FIND.account.address == 0x097bafa4e0b48eef {
+		// This is for mainnet
+			if merchAccount != 0x097bafa4e0b48eef {
```
Comment (bjartek) [inline contracts/FIND.cdc:1878]: These merch addresses needs to be changed once we have them, but this is good for now

File: find_dapper_test.go [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+			WithArg("receiver", "user2"),
+		).
+			AssertSuccess(t).
+			AssertEvent(t, "A.f8d6e0586b0a20c7.FIND.Moved", map[string]interface{}{
+				"name": "user1",
+			})
+
+	})
+
+	t.Run("Should automaticall set Find ame to empty if sender have none", func(t *testing.T) {
```
Comment (bjartek) [inline find_dapper_test.go:138]: this test has some typos

File: find_dapper_test.go [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+			WithArg("merchAccount", "find"),
+			WithArg("name", user),
+			WithArg("addon", "A.f8d6e0586b0a20c7.Dandy.NFT"),
+			WithArg("amount", 10.0),
+		).
+			AssertFailure(t, "This addon is not available.")
+
+	})
+
+	// Should not be able to manage fund
```
Comment (bjartek) [inline find_dapper_test.go:335]: can we just remove this if it is not relevant anymore?

File: transactions/buyleaseForSaleDapper.cdc [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import FindMarket from "../contracts/FindMarket.cdc"
+import FIND from "../contracts/FIND.cdc"
+import Profile from "../contracts/Profile.cdc"
+import DapperUtilityCoin from "../contracts/standard/DapperUtilityCoin.cdc"
+import FindLeaseMarketSale from "../contracts/FindLeaseMarketSale.cdc"
+import FindLeaseMarket from "../contracts/FindLeaseMarket.cdc"
+
+//TODO: test, and rename to Dapper, repeat for other tx
+transaction(dapperAddress: Address, leaseName: String, amount: UFix64) {
```
Comment (bjartek) [inline transactions/buyleaseForSaleDapper.cdc:9]: this should be called merchantAddress or?

**Review by bjartek — COMMENTED** [review]
We also need to create the corresponding scripts to all "buy" type transactions that have the same parameters as the buy transaction and returns this struct

pub struct PurchaseData {
    pub let id: UInt64
    pub let name: String
    pub let amount: UFix64
    pub let description: String
    pub let imageURL: String

    init(id: UInt64, name: String, amount: UFix64, description: String, imageURL: String) {
        self.id = id
        self.name = name
        self.amount = amount
        self.description = description
        self.imageURL = imageURL
    }
}

File: find_dapper_test.go (hunk 7 above)
Comment (Bamthelearner) [reply find_dapper_test.go]: removed

File: transactions/buyleaseForSaleDapper.cdc (hunk 8 above)
Comment (Bamthelearner) [reply transactions/buyleaseForSaleDapper.cdc]: This is not needed when I look again at it. 
I just deleted that

**Pull request merged**